src/pricecypher/storage/file_storage.py

```diff
@@ -0,0 +1,114 @@
+        :param path_remote_prefix: Location within / on top of `path_remote_base` where remote artifacts will end up.
+
+        NB: Values should be such that a locally stored 'file' at `path_local / file` matches with a remote copy at
+        `path_remote_base / path_remote_prefix / file`.
+        """
+        self._path_local = path_local
+        self._path_remote_base = path_remote_base
+        self._path_remote_prefix = path_remote_prefix
+
+    def get_path_local(self, filename: str) -> str:
+        return os.path.join(self._path_local, filename)
+
+    def get_path_remote(self, filename: str, full: bool = True) -> str:
+        """
+        :param filename: Name of the file.
+        :param full: Whether the full remote path, so including `self._path_remote_base`, should be returned.
+        :return: Path in remote storage. Either absolute (including 'base') or relative, depending on given `full`.
+        """
+        suffix = os.path.join(self._path_remote_prefix, filename)
+
+        if not full:
+            return suffix
+
+        return os.path.join(self._path_remote_base, suffix)
+
+    @classmethod
+    def get_scheme(cls, uri_as_string):
+        uri = smart_open.parse_uri(uri_as_string)
+        return uri.scheme if hasattr(uri, 'scheme') else None
+
+    def get_scheme_local(self, filename: str) -> str:
+        return self.get_scheme(self.get_path_local(filename))
+
+    def get_scheme_remote(self, filename: str) -> str:
+        return self.get_scheme(self.get_path_remote(filename))
+
+    @contextmanager
+    def save(self, filename: str, mode: str = 'w') -> str:
+        """
+        Open a file handler context to a new file in local storage.
+        NB: the locally saved file should be uploaded to the remote storage automatically / externally.
+
+        :param filename: Name of the file to save.
+        :param mode: (Optional) Mimicks the `mode` parameter of the built-in `open` function. Defaults to 'w'.
+
+        See Also
+        --------
+        - `Standard library reference <https://docs.python.org/3.7/library/functions.html#open>`__
+        - `smart_open README.rst <https://github.com/RaRe-Technologies/smart_open/blob/master/README.rst>`__
+        """
+        if not self._path_local:
+            raise InvalidStateException("The `path_local_outputs` and `path_local_outputs` must be set to save files.")
+
+        local = self.get_path_local(filename)
+
+        logging.info(f"Saving file, local path = '{local}', remote path = '{self.get_path_remote(filename)}'...")
+
+        if self.get_scheme(local) == 'file':
+            logging.debug("Making non-existing directories on the path to parent of `file_path_local`...")
+            Path(local).parent.mkdir(parents=True, exist_ok=True)
+
+        with smart_open.open(local, mode) as file:
+            yield file
+
+    @contextmanager
+    def load(self, path: Union[Path, str], mode: str = 'r') -> str:
```

Comment:
I assume these functions are only executed locally by the datascience team? If not, have we thought about possible RCEs?

Reply:
Currenty, we may assume that all scripts being executed are in our (1st party) full control indeed. Allowing others to execute scripts would, in itself by definition, already allow for remote code execution, right?

Reply:
In fact the whole system works with RCE as one of the features, fair enough.